File: generic-code-challenges/minesweeper-mine-counting/minesweeper.js
let MINEFIELD = [
  [null, null, null, null, 'mine'],
  [null, null, null, null, 'mine'],
  ['mine', null, null, null, null],
  [null, null, null, 'mine', null],
  [null, 'mine', null, null, null]
]

// bounds-checking helper function
// returns an "empty" value if the coordinate doesn't exist.
function getCell(grid, row, col) {
  let empty = null;
  if (row < 0 || col < 0) {
    return empty;
  } else if (row >= grid.length) {
    return empty;
  } else if (col >= grid[row].length) {
    return empty;
  } else {
    return grid[row][col]
  }
}

// this function accepts a 2D array representing
// a minesweeper minefield. It returns a new 2D
// array where each cell has a number representing
// how many mines that cell touches, 
// let MINEFIELD = [
//   [null, null, null, null, 'mine']
//   [null, null, null, null, 'mine']
//   ['mine', null, null, null, null]
//   [null, null, null, 'mine', null]
//   [null, 'mine', null, null, null]
// ]
// [
//  [0, 0, 0, 2, *]
//  [1, 1, 0, 2, *]
//  [*, 1, 1, 2, 2]
//  [2, 2, 2, *, 1]
//  [1, *, 2, 1, 1]
// ]
function markMines(field) {
  for (let row = 0; row < field.length; row++) {
    for (let col = 0; col < field[row].length; col++) {
      field[row][col] = countMines(field, row, col)
    }
  }
  
  return field
}

function countMines(field, row, col) {
  // we must check for both 'mine' and 9 since we're modifying
  // the original array.
  if (field[row][col] === 'mine' || field[row][col] === 9) {
    return 9;
  }
  let result = 0;
  
  if (getCell(field, row - 1, col) === 'mine') { // north
    result++
  }
  if (getCell(field, row + 1, col) === 'mine') { // south
    result++
  }
  if (getCell(field, row, col + 1) === 'mine') { // east
    result++
  }
  if (getCell(field, row, col - 1) === 'mine') { // west
    result++
  }
  
  if (getCell(field, row - 1, col - 1) === 'mine') { // nw
    result++
  }
  if (getCell(field, row - 1, col + 1) === 'mine') { // ne
    result++
  }
  if (getCell(field, row + 1, col - 1) === 'mine') { // sw
    result++
  }
  if (getCell(field, row + 1, col + 1) === 'mine') { // se
    result++
  }
  
  return result;
}

console.log('Minefield:')
console.log(MINEFIELD)
console.log()

let result = markMines(MINEFIELD)

console.log('Result:')
console.log(result)
console.log()

simpleAssert(result[0][0], 0)
simpleAssert(result[1][0], 1)
simpleAssert(result[1][1], 1)
simpleAssert(result[3][0], 2)
simpleAssert(result[4][4], 1)

function simpleAssert(actual, expected) {
  if (expected !== actual) {
    console.log("expected:", expected, "actual:", actual)
  }
}

File: generic-code-challenges/minesweeper-mine-counting/Minesweeper.java
public class Minesweeper {
    private static  int[][] MINEFIELD = {
        {0, 0, 0, 0, 9},
        {0, 0, 0, 0, 9},
        {9, 0, 0, 0, 0},
        {0, 0, 0, 9, 0},
        {0, 9, 0, 0, 0},
    };

    public static void print2DArray(int[][] grid) {
        for (int row = 0; row < grid.length; row++) {
            for (int col = 0; col < grid.length; col++) {
                System.out.print(grid[row][col] + " ");
            }
            System.out.println();
        }

    }

    public static void main(String[] args) {
        System.out.println("Original:");
        print2DArray(MINEFIELD);
        System.out.println();

        markMines(MINEFIELD);

        System.out.println("Result:");
        print2DArray(MINEFIELD);
        System.out.println();
    }

    public static int[][] markMines(int[][] mineField) {
        for (int row = 0; row < mineField.length; row++) {
            for (int col = 0; col < mineField.length; col++) {
                mineField[row][col] = countMines(mineField, row, col);
            }
        }

        return mineField;
    }

    private static int getCell(int[][] grid, int row, int col) {
        if (row < 0 || col < 0) {
            return 0;
        } else if (row >= grid.length) {
            return 0;
        } else if (col >= grid[row].length) {
            return 0;
        } else {
            return grid[row][col];
        }
    }

    private static int countMines(int[][] field, int row, int col) {
        int cell = getCell(field, row, col);
        if (cell == 9) {
            return 9;
        } else {
            int nearbyMines = 0;
            int[][] neighborDirections = {
                    {-1, 0}, {1, 0}, {0, 1}, {0, -1},  // north, south, east and west
                    {-1, -1}, {-1, 1}, {1, 1}, {1, -1} // nw, ne, se, sw
            };

            for (int i = 0; i < neighborDirections.length; i++) {
                int[] offsets = neighborDirections[i];

                int rowIndex = row + offsets[0];
                int colIndex = col + offsets[1];
                if (getCell(field, rowIndex, colIndex) == 9) {
                    nearbyMines++;
                }
            }
            return nearbyMines;
        }
    }
}